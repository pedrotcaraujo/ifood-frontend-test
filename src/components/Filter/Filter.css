.Filter {
    display: flex;
    margin: 0 30px;
    flex-wrap: wrap;
    justify-content: center;
}