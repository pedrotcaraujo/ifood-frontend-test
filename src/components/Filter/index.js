import React, { Fragment, Component } from 'react';
import axios from 'axios';

const URL = 'http://www.mocky.io/v2/5a25fade2e0000213aa90776'

class Filter extends Component {
    constructor() {
        super()
        this.state = { data: null };
    }
    componentDidMount() {
        axios.get(URL)
            .then(({ data }) => this.setState(data))
            .catch(err => console.log(err));
    }

    buildFilters(filter) {
        if (filter.id === 'locale') {
            return (
                <select>
                    {filter.values.map(({value, name}) => <option key={value} value={value}>{name}</option>)}
                </select>
            )
        }

        if (filter.id === 'country') {
            return (
                <select>
                    {filter.values.map(({value, name}) => <option key={value} value={value}>{name}</option>)}
                </select>
            )
        }

        if (filter.id === 'timestamp') {
            return (
                <input type="date"/>
            )
        }

        if (filter.id === 'limit') {
            return (
                <input type="number" max="50" min="20"/>
            )
        }

        if (filter.id === 'offset') {
            return (
                <input type="number"/>
            )
        }
    }
    render() {
        return (
            <Fragment>
                {this.state.filters ? (
                    <div>
                        {this.state.filters.map(this.buildFilters)}
                    </div>
                ) : 'Loading...'}
            </Fragment>
        )
    }
}

export default Filter;