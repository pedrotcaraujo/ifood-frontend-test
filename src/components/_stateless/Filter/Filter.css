.Filter {
    display: flex;
    margin: 0 30px;
    flex-wrap: nowrap;
    justify-content: center;
}

@media all and (max-width: 736px) {
    .Filter {
        flex-wrap: wrap;
    }
}