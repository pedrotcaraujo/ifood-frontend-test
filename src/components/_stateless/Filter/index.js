import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import { Dropdown, Input } from '../Fields';
import './Filter.css';

import debounce from '../../../utils/debounce';

const DEBOUNCE_TIMOUT = 1000;
const FILTER = {}

const supportedFilters = {
    locale: (data, onChange) => renderDropdown(data, onChange),
    country: (data, onChange) => renderDropdown(data, onChange),
    limit: (data, onChange) => (
        <Input key={data.id} 
            inputProps={{
                placeholder: `${data.name} min: ${data.validation.min} max: ${data.validation.max}`,
                name: data.id,
                type: 'number',
                max: data.validation.max,
                min: data.validation.min,
                onChange: (evt) => _handleChange(evt, onChange)
            }}
        />),
    offset: (data, onChange) => (
        <Input key={data.id} 
            inputProps={{
                placeholder: data.name,
                name: data.id,
                type: 'number',
                onChange: (evt) => _handleChange(evt, onChange)

            }}
         />
    )
};

const renderDropdown = (data, onChange) => (
    <Dropdown key={data.id} 
        dropdownProps={{ 
            name: data.id, 
            onChange: (evt) => _handleChange(evt, onChange) 
        }} 
        options={[{name: data.name, value: ''}, ...data.values]} 
/>);


// Render the filter fields by reflection
const renderFilter = (filter, onChange) => typeof supportedFilters[filter.id] === 'function' && supportedFilters[filter.id](filter, onChange)


// Handle all filters and send by props onChange
const _handleChange = (evt, action) => {
    evt.persist();
    debounce(({ target }, action) => {
        const current = {
            name: target.name,
            value: target.value
        }

        FILTER[current.name] = current.value;
        FILTER.timestamp = moment().format();

        action({ current, data: FILTER })
    }, DEBOUNCE_TIMOUT)(evt, action);
}

const Filter = ({data, onChange}) => (
    <div className="Filter">
        <Input 
            inputProps={{
                placeholder: "Buscar por nome",
                name: 'search',
                type: 'text',
                onChange: (evt) => _handleChange(evt, onChange)
            }}
        />
        {data.filters.map((filter) => renderFilter(filter, onChange))}
    </div>
)

Filter.propTypes = {
    data: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired
};

export default Filter;