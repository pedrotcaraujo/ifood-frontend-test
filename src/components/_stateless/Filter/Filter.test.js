import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Filter from './index';

const onChangeMock = jest.fn();

it('renders correctly', () => {
  const tree = renderer
    .create(<Filter data={{ filters: [] }} onChange={onChangeMock} />)
    .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Filter data={{ filters: [] }} onChange={onChangeMock} />, div);
});