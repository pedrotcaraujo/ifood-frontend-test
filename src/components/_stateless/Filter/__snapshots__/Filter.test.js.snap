// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Filter"
>
  <div
    className="Input"
  >
    <input
      className="Input-field"
      name="search"
      onChange={[Function]}
      placeholder="Buscar por nome"
      type="text"
    />
  </div>
</div>
`;
